[Clang] Add a builtin that deduplicate types into a pack

ilya-biryukov opened the pull request:
The new builtin `__builtin_dedup_pack` removes duplicates from list of types.

The added builtin is special in that they produce an unexpanded pack
in the spirit of P3115R0 proposal.

Produced packs can be used directly in template argument lists and get
immediately expanded as soon as results of the computation are
available.

It allows to easily combine them, e.g.:

```cpp
template <class ...T>
struct Normalize {
  // Note: sort is not included in this PR, it illustrates the idea.
  using result = std::tuple<
    __builtin_sort_pack<
      __builtin_dedup_pack<int, double, T...>...
    >...>;
}
;
```

Limitations:
- only supported in template arguments and bases,
- can only be used inside the templates, even if non-dependent,
- the builtins cannot be assigned to template template parameters.

The actual implementation proceeds as follows:
- When the compiler encounters a `__builtin_dedup_pack` or other type-producing
  builtin with dependent arguments, it creates a dependent
  `TemplateSpecializationType`.
- During substitution, if the template arguments are non-dependent, we
  will produce: a new type `SubstBuiltinTemplatePackType`, which stores
  an argument pack that needs to be substituted. This type is similar to
  the existing `SubstTemplateParmPack` in that it carries the argument
  pack that needs to be expanded further. The relevant code is shared.
- On top of that, Clang also wraps the resulting type into
  `TemplateSpecializationType`, but this time only as a sugar.
- To actually expand those packs, we collect the produced
  `SubstBuiltinTemplatePackType` inside `CollectUnexpandedPacks`.
  Because we know the size of the produces packs only after the initial
  substitution, places that do the actual expansion will need to have a
  second run over the substituted type to finalize the expansions (in
  this patch we only support this for template arguments, see
  `ExpandTemplateArgument`).

If the expansion are requested in the places we do not currently
support, we will produce an error.

More follow-up work will be needed to fully shape this:
- adding the builtin that sorts types,
- remove the restrictions for expansions,
- implementing P3115R0 (scheduled for C++29, see
  https://github.com/cplusplus/papers/issues/2300).

ilya-biryukov committed:
[Clang] Add a builtins that deduplicate types into a pack

The new builtin `__builtin_dedup_pack` removes duplicates from list of types.

The added builtin is special in that they produce an unexpanded pack
in the spirit of P3115R0 proposal.

Produced packs can be used directly in template argument lists and get
immediately expanded as soon as results of the computation are
available.

It allows to easily combine them, e.g.:

```cpp
template <class ...T>
struct Normalize {
  // Note: sort is not included in this PR, it illustrates the idea.
  using result = std::tuple<
    __builtin_sort_pack<
      __builtin_dedup_pack<int, double, T...>...
    >...>;
}
;
```

Limitations:
- only supported in template arguments and bases,
- can only be used inside the templates, even if non-dependent,
- the builtins cannot be assigned to template template parameters.

The actual implementation proceeds as follows:
- When the compiler encounters a `__builtin_dedup_pack` or other type-producing
  builtin with dependent arguments, it creates a dependent
  `TemplateSpecializationType`.
- During substitution, if the template arguments are non-dependent, we
  will produce: a new type `SubstBuiltinTemplatePackType`, which stores
  an argument pack that needs to be substituted. This type is similar to
  the existing `SubstTemplateParmPack` in that it carries the argument
  pack that needs to be expanded further. The relevant code is shared.
- On top of that, Clang also wraps the resulting type into
  `TemplateSpecializationType`, but this time only as a sugar.
- To actually expand those packs, we collect the produced
  `SubstBuiltinTemplatePackType` inside `CollectUnexpandedPacks`.
  Because we know the size of the produces packs only after the initial
  substitution, places that do the actual expansion will need to have a
  second run over the substituted type to finalize the expansions (in
  this patch we only support this for template arguments, see
  `ExpandTemplateArgument`).

If the expansion are requested in the places we do not currently
support, we will produce an error.

More follow-up work will be needed to fully shape this:
- adding the builtin that sorts types,
- remove the restrictions for expansions,
- implementing P3115R0 (scheduled for C++29, see
  https://github.com/cplusplus/papers/issues/2300).

ilya-biryukov commented:
RFC for this change: https://discourse.llvm.org/t/rfc-adding-builtin-for-deduplicating-type-lists/80986

ilya-biryukov commented:
I still need to add code that preserves source locations for template arguments and tests for it, but otherwise this is ready for review and I would like to get some initial feedback.

The RFC did not get any comments so far, but also happy to discuss the general design there.

usx95 reviewed (commented):
We also need documentation explaining the new builtin.
Can you also add ReleaseNotes.

ilya-biryukov commented on clang/include/clang/Basic/Builtins.h:
```diff
@@ -309,7 +309,10 @@ enum BuiltinTemplateKind : int {
+  BTK__type_pack_element,
+
+  /// This names the __type_list_dedup BuiltinTemplateDecl.
+  BTK__type_list_dedup,
```
On the argument of adding `VariadicTransformType`, maybe it's useful for something, but I definitely don't see why introducing a new bulitin template is more complicated than introducing a new type trait. I think they're quite on par in terms of complexity and we should be looking at what fits the language design better (see first two bullet points above on why I feel it's a better choice for the language).

zygoloid replied:
Given that we already have builtin templates, if we're concerned about the boilerplate involved in adding a new one, perhaps we could remove most of that boilerplate instead? It looks like an xmacro (`.def` file) approach could handle most of these changes. Though I'd suggest we decouple that refactoring from this change.

Unless we think that builtin templates were a mistake, and we instead would prefer to deprecate and eventually remove our support for builtin templates?

For what it's worth, as I recall the idea was that builtin templates would be a lot *less* invasive than builtin type traits for this kind of thing. They don't need a new keyword, separate parsing rules, a new AST representation, etc -- they're just another kind of type template name. And they support type arguments, non-type arguments, and templates as arguments without extra work. And they can reuse our normal type-checking for templates. The big downside is that they don't work in C, but for things intended for operating on parameter packs such as the existing two builtin templates and this new one, that's fine.

philnik777 replied:
FYI I have a local patch to refactor most of the code into just a `BuiltinTemplates.td`. In the end that should only make it required to modify `BuiltinTemplate.td` and `SemaTemplate.cpp`.

ilya-biryukov replied:
> Unless we think that builtin templates were a mistake, and we instead would prefer to deprecate and eventually remove our support for builtin templates?

I certainly feel they are a useful tool to have and we should probably keep them.
I'm definitely very supportive of reducing the boilerplate.

On that front, @philnik777 do you feel like you digging out that local patch or should I take a stab at it?
In any case, I'd do this in a separate change. The boilerplate present in this commit is still very much manageable (and already done), so I don't think that blocking this PR on removing this boileerplate would be my first choice.

mizvekov replied:
One downside of builtin templates is that they can affect mangling, if we wish to support cases where we can't specialize them until dependencies are resolved.

For example, we can't do anything with __make_integer_seq if the count is value dependent, which means in order to support that case, we build a canonical TST which is named by the builtin template name, and this means `__make_integer_seq` will be mangled.

cor3ntin commented:
(This is in my list of PR/RFC to review, I'll try to get to it this week)

ilya-biryukov commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -38,9 +40,11 @@
 #include "llvm/ADT/SmallBitVector.h"
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/raw_ostream.h"
```
thanks! removed it.

zygoloid commented on clang/include/clang/Basic/Builtins.h:
```diff
@@ -309,7 +309,10 @@ enum BuiltinTemplateKind : int {
-  BTK__type_pack_element
+  BTK__type_pack_element,
+
+  /// This names the __type_list_dedup BuiltinTemplateDecl.
```
This name seems inconsistent with the previous builtin: using the `__type_pack` prefix for both would be more uniform.

philnik777 replied:
I think we've decided somewhere that all new builtin functionality should have a `__builtin_` prefix. The main reason was to avoid naming conflicts. While I don't think that's a huge issue here, we should probably be consistent in that regard. So this should probably be `__builtin_type_pack_dedup`.

ilya-biryukov replied:
Thanks, I don't actually know why I didn't pick `type_pack` prefix in the first place. I think I was confused when I started and thought that `type_pack_element` was more different than it ended up being.

It's now `__builtin_type_pack_dedup`, as suggested.

zygoloid commented on clang/test/SemaTemplate/temp-param-list-dedup.cpp:
```diff
@@ -0,0 +1,57 @@
+// RUN: %clang_cc1 %s -verify
```
I'd like to also see some testing of the case where the template arguments have different sugar but are canonically the same, and when they differ by only cv-qualifiers. Those seem like likely cases for us to get wrong. Eg:

```c++
using T = int;
using CT = const T;
static_assert(__is_same(__type_list_dedup<TypeList, T, int, const int, const T, CT, const CT>,
                        TypeList<int, const int>));
```

ilya-biryukov replied:
Done. I have also added a few tests with arrays and pointers, which should also exhibit a few more combinations of types that subsume qualifiers and could be different when canonical/non-canonical.

zygoloid commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3158,6 +3161,33 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+    assert(Template.getKind() == clang::TemplateArgument::Template);
+    assert(Ts.getKind() == clang::TemplateArgument::Pack);
+    TemplateArgumentListInfo SyntheticTemplateArgs;
+    llvm::DenseSet<QualType> Seen;
```
Consider a `SmallDenseSet` here to avoid heap allocation for shorter lists.

ilya-biryukov replied:
Done.

zygoloid reviewed (commented):
This approach makes sense to me. I'd intended that `__type_pack_element` would be the start of a family of builtin templates operating on packs, but sadly didn't get very far down that road.

ilya-biryukov commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3158,6 +3161,33 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+      SyntheticTemplateArgs.addArgument(TemplateArgumentLoc(
+          TemplateArgument(T), SemaRef.Context.getTrivialTypeSourceInfo(
+                                   T.getAsType(),
+                                   /*FIXME: add location*/ SourceLocation())));
```
After looking at the code, I am afraid the not-very-precise source location will stay a limitation of this builtin (it also seems like it's the case for some other cases involving packs? as soon as a template argument gets added into a pack, it seems we should be loosing its location information at least in some cases)

We do not seem to be tracking the `TemplateArgumentLoc`s of `TempateArgument`s we get from the pack very precisely. While it's possible to poke at the `TemplateArgumentList` provided into the function and match them with the types I get from `Converted`, I am not entirely sure what the structure of the `TemplateArgumentListInfo` is [1] and whether the complexity around it is worth it?

@zygoloid any thoughts on this? is it worth digging into the various representations the passed `TemplateArgumentListInfo` might have or is it a dead end? 

<hr>
[1]: is it always flattened? does it always have packs? can it be both?

zygoloid replied:
I think the only thing we would use this source location for is to diagnose in `CheckTemplateIdType` if the template arguments aren't suitable for use as arguments of `Template`, which we already checked before getting here. So I don't think it's worth going to any effort to get a good location. Let's just pass in `TemplateLoc` as the location.

(It's notable that `__make_integer_seq` uses (eg) `TemplateArguments[1].getLocation()`. We could do the same, but that will break once we add universal template parameters, which [seem to be making good progress through WG21](https://github.com/cplusplus/papers/issues/1662) -- where there might be only a single template argument that pack-expands to all the arguments of `__make_integer_seq`, and likewise here. So maybe it'd be worth changing `__make_integer_seq` to pass in `TemplateLoc` instead while we're thinking about this.)

ilya-biryukov commented:
Friendly ping for another round of review.

mizvekov reviewed (commented):
Thanks for working on this problem!

I think the current approach is limited in ABI stability, we could broaden the usability here by designing this builtin so it produces a type list which is unique per set of types (order independent).

I had a chat with @lewissbaker a couple of months ago, and they need a similar facility for stdexec, except that ABI stability is important for them.

One possible approach would be to sort the types according to their Itanium mangling, or something similarly stable.

Would that suit your needs as well?

mizvekov commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3158,6 +3161,33 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+                                    "a template and a parameter pack");
+    TemplateArgument Template = Converted[0];
+    TemplateArgument Ts = Converted[1];
+    if (Template.isDependent() || Ts.isDependent())
```
I think the only reason we would need to hold off on building the template specialization would be if Ts contains an unexpanded pack, otherwise we are good to go, subsequent substitutions will produce the correct type list anyway.

This would also reduce the size of the argument list.

Be aware that holding this specialization will make this builtin appear in mangling, so I'd avoid doing that unless necessary.

ilya-biryukov replied:
We could remove the dependency on the template, but I don't think we can ignore the dependency inside the types themselves.
Consider
```cpp
template <class T, class U, template <class...> class TL = TypeList>
struct Foo {
  // 1. the substitution has to be postponed until T and U are known.
  using result = __dedup<TypeList, T, U>;
  // 2. this is already computed right away.
  using result2 = __dedup<TypeList, int, int>; 
  // 3. this **could** be computed right away, but it is not now.
  using result3 = __dedup<TL, int, int>
};

static_assert(__is_same(TypeList<int>, Foo<int, int>::result));
static_assert(__is_same(TypeList<int, doubel>, Foo<int, double>::result));
```

- Is there something I'm missing that would allow (1) to be substituted earlier?
- Do you feel strongly that we need to do (3) ?

> Be aware that holding this specialization will make this builtin appear in mangling, so I'd avoid doing that unless necessary.

I do not think we can avoid this completely (see above) and it's not unprecedented: other builtin templates can already appear in mangling too when the arguments are dependent. So I do think it's necessary and I'm not sure which problems it carries.

mizvekov replied:
You are right that we can't drop out the builtin template specialization just yet if any of the arguments are dependent.

But we can still unique the list in that case. This actually even applies for unexpanded packs as well.

One advantage I see is that you would avoid the possibility of the user introducing multiple overloads that are equivalent, which you wouldn't be able to partial order later.

Ie

```C++
template <class...> struct A;
template <class T> void f(__builtin_type_pack_dedup<A, T>) {}
template <class T> void f(__builtin_type_pack_dedup<A, T, T>) {}
```
This would be ill-formed because `f` is being redefined.

This also increases the ABI stability for the same reason. If the user introduces another repeated template argument, it's still results in the same templated entity with the same mangling.

> Do you feel strongly that we need to do (3) ?

See above.

> I do not think we can avoid this completely (see above) and it's not unprecedented: other builtin templates can already appear in mangling too when the arguments are dependent. So I do think it's necessary and I'm not sure which problems it carries.

Right, it's not completely unavoidable, but at least for the dependent template name case that we can resolve early, it makes it so this builtin can be introduced without causing an ABI break.

cor3ntin reviewed (changes requested):
There are some comments on the RFC, fyi

ilya-biryukov commented:
I have also left a few comments in the parallel discussion on the RFC.

ilya-biryukov commented:
@cor3ntin and everyone else, would you mind taking another look?

I wanted to bring this up again and tried to do all the proposed changes:
- we now have a builtin to sort types by mangled name.
- the builtins now expand directly into a list of template arguments.
- the builtins are now named `__builtin_dedup_types` and `__builtin_sort_types` as it's grown on me that `type_pack` is actually not the best name (even if we have some existing ones called like this).

I was worried that the latter would be too complicated, but it actually wasn't.
I had to make sure that it's only used in template arguments, but that was fairly easy (I hope I did not miss any important cases).

I would appreciate if folks took another look here and provided their feedback.
If this is going in the right direction, I'll write up the release notes and do any final changes that are necessary.

cor3ntin commented:
Sorry for the delay, I lefts some comments here 
https://discourse.llvm.org/t/rfc-adding-builtin-for-deduplicating-type-lists/80986/25?u=cor3ntin

ilya-biryukov commented:
I have made quite a bit of progress here, but we also need the deduplication builting urgently to help ease the pain for
some of the big compilations that started hitting our infrastructure limits internally.

I have created #139730 with the initial and simple version of the builtin that I would really love to land ASAP as I expect this PR to take more time to cook (see open questions below, I'm sure there will be a few more comments). I fully understand the desire to have a better version of the builtin in the long term and I am happy to continue working on this until it lands. But we also need the build time savings at a shorter timeframe, so I hope that having a simpler version in the meantime would not be  a particularly contentious point. Let me know, though. 

With that out of the way, @cor3ntin could you take a look at the new version here?
It still has a few rough edges, but I want to make sure the choices made here make sense to you, in particular:
- using `TemplateSpecializationType` for dependent instantiations and for sugar after we have a template pack after substitution,
- doing a second run during template instantiations when we need to expand the packs that are delayed,
- a new type (that shares a common base class with `SubstTemplateParmPackType`) to represent the pack after substitution.

Also eager to hear feedback from others, just calling out @cor3ntin as someone who has been the author of the relevant standard proposals that this PR is trying to mimic.

ilya-biryukov commented:
Note to people reviewing this: please ensure you "hide whitespace" via `git diff -w` or the corresponding setting on GitHub. For your own sanity when looking at changes in `TreeTransform.h`, which are otherwise really complicated to validate. I've deliberately kept the code at the same place to have reasonable diffs.

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+    llvm::SmallVector<TemplateArgument> OutArgs;
+    llvm::SmallDenseSet<QualType> Seen;
+    // Synthesize a new template argument list, removing duplicates.
+    for (auto T : Ts.getPackAsArray()) {
```
For this array, should we just do an insert into a MapVector instead?

ilya-biryukov replied:
I couldn't make it work. The defaults gives me `SmallVector<pair<QualType, TemplateArgument>>`, and if I specify `MapVector<QualType, TemplateArgument, DenseMap<QualType, unsigned>, SmallVector<TemplateArgument>>`, `try_emplace` fails because it still tries to construct a pair inside the vector.

Am I holding it wrong?

erichkeane replied:
I suspect it is a 'holding it wrong'?  Though I probably MEANT to suggest `SetVector`?  Does that make it easier?

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+    if (Converted[0].isDependent())
+      return QualType();
+
+    auto InputArgs = Converted[0].getPackAsArray();
```
Don't use auto here.

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+        SemaRef.getASTContext().createMangleContext());
+
+    // Prepare our sort keys, i.e. the mangled names.
+    llvm::SmallVector<std::string> MangledNames(InputArgs.size());
```
Any reason we can't use a range-for with a zip here or something?

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+    assert(Converted.size() == 1);
+    assert(Converted[0].getKind() == TemplateArgument::Pack);
+    // Delay if we have any dependencies, the mangled names may change after
+    // subsistution or may not be well-defined for dependent types.
```
```suggestion
    // substitution or may not be well-defined for dependent types.
```

erichkeane commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+      return MangledNames[L] < MangledNames[R];
+    });
+
+    llvm::SmallVector<TemplateArgument> SortedArguments;
```
This whole part of this function (going through the list so many times) feels pretty convoluted/unnecessary, particularly with how many iteration-based things there are.  IMO, we should just be able to create an object of each type, then sort it correctly, then std::transform it into the sorted args.

erichkeane commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -65,6 +72,39 @@ class CollectUnexpandedParameterPacksVisitor
     }
 
+    bool addUnexpanded(const SubstBuiltinTemplatePackType *T) {
+      // TODO: source location
```
TODO?

erichkeane reviewed (commented):
Thanks for the ping @cor3ntin.  This is more or less on the right path.  I wish we could break this up a bit to make reviewing more efficient, but I think this is vastly better than any temporary solution.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
     return HasNoTypeMember;
   }
+  case BTK__builtin_dedup_pack: {
+    assert(Converted.size() == 1 && "__builtin_dedup_pack should be given "
+                                    "a parameter pack");
```
Should that not be a diagnostic?

ilya-biryukov replied:
Since we require the arguments to match first, I don't think we would ever get anything other than a pack here. Other builtins follow the same pattern. Am I wrong here? Could you provide an example where this happens?

erichkeane replied:
IIRC, this should be caught by SemaChecking based on the builtin definition.  See the change to `[clang/include/clang/Basic/BuiltinTemplates.td](https://github.com/llvm/llvm-project/pull/106730/files#diff-d3763371ff5a25641558b8e3a2d166846d2dbfc8738da3f37f80b49a397b1f9d)`, so normal template-function-call checking should take care of these.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -3331,6 +3343,65 @@ checkBuiltinTemplateIdType(Sema &SemaRef, BuiltinTemplateDecl *BTD,
+        TemplateArgument::CreatePackCopy(Context, OutArgs));
+  }
+  case BTK__builtin_sort_pack: {
+    assert(Converted.size() == 1);
```
Ditto, diagnostics?

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1666,6 +1685,21 @@ namespace {
+      if (!getSema().ArgPackSubstIndex || !T->isSugared() ||
+          !isPackProducingBuiltinTemplateName(T->getTemplateName()))
+        return TreeTransform::TransformTemplateSpecializationType(TLB, TL);
+      // Look through sugar to get to the SubstBuiltinTemplatePackType that we
+      // need to substitute into.
+      QualType R = TransformType(T->desugar());
```
How does a type sugar impede substitution?

mizvekov replied:
Yes the default TreeTransform implementation will rerun semantic actions, so its a bit inefficient for sugar-only transforms, but you can customize that by providing overrides for the Rebuild* functions.

The issue I see in this case here is that you are just dropping the TemplateSpecializationType, why can't you just build a new one here to wrap over the transformed underlying type?

ilya-biryukov replied:
> The issue I see in this case here is that you are just dropping the TemplateSpecializationType, why can't you just build a new one here to wrap over the transformed underlying type?

`TemplateSpecializationType` looks like the right sugar for the whole pack, but after we pick a single element from it I don't think it models what we want correctly. I.e. `dedup<int, double> == argpack{int, double}`, but **not equal** to the individual `int` or `double` after we've picked them.
So I decided to stick with the lack of sugar rather than pick the incorrect one (from that perspective).

Or do you feel `TemplateSpecializationType` is a good fit here?

mizvekov replied:
Yeah we need a representation for how the type was written, we try to avoid faking resolved types as written ones.

We model all current builtin templates as type aliases, and I still see this as a good fit here.

When unresolved (due to dependencies), they stay as canonical template specialization types, and when they  are resolved they simply become sugar by turning into aliases over whatever the builtin template specialization resolves to.

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -5054,95 +5089,128 @@ bool TreeTransform<Derived>::TransformTemplateArguments(
+      return true;
+    SubstIndex.reset();
+
+    // Some packs only know their lengths after substitution. We might be able
+    // to expand those here.
+    if (TryExpandingSubstPacks &&
+        OutPattern.getArgument().containsUnexpandedParameterPack()) {
```
Thanks for the explanation - I'll pull the code and get myself familar with the fancy builtin so I might not respond very quickly. Please feel free to go ahead once others have approved

Just disregard my "request changes" flag, my mouse slipped when I submitted comments :)

zyn0217 replied:
So this branch is used when rebuilding a substituted `dedup` node. E.g. for `Ts = <int, int, int, float>`, we're, at this point, having `dedup<int, int, int, float>...` plus a canonical `SubstBuiltinTemplatePackType` which is used to mark it as an unexpanded pack.

This feels a bit peculiar to me because we're trying to inline node-specific logic (?) into what's supposed to be a generic expansion handler.

zyn0217 replied:
I'm thinking if we shall move the second-pass-expansion-transform out of TransformTemplateArguments. It seemed that so far only TST that model a "synthesized" pack type can benefit from the two level transform, and separating that explicitly makes the structure clearer to me.

Maybe we could move it to just after the first transform of template arguments inside TST, or possibly override the rebuild process for TST itself? Essentially, we're now just reusing TST to represent the novel "synthesized" packs (I haven't come up with a better name yet), and ideally we want to extend it / probably invent a new type node to support other pack forms outside of the template context, so not baking it into TemplateArgument transform makes some sense to me.

That said, I'm not sure what direction we want to take here, so CC @cor3ntin @erichkeane @mizvekov for ideas!

zyn0217 commented on clang/include/clang/AST/Type.h:
```diff
@@ -6484,6 +6499,57 @@ class SubstTemplateTypeParmType final
+  }
+};
+
+/// Represents the result of substituting a builtin template as a pack.
+class SubstBuiltinTemplatePackType : public SubstPackType {
+  friend class ASTContext;
+
+  SubstBuiltinTemplatePackType(QualType Canon, const TemplateArgument &ArgPack);
```
Can you explain what is preventing us from reusing `SubstTemplateTypeParmPackType`?

The commit message said they are similar, but didn't explain why.

One possible reason I can think of is that, `collectUnexpandedParameterPacks` doesn't like `SubstTemplateTypeParmPackType` node because some clients don't expect it (e.g. in type deduction). But that can always be distincted (inside the collector) by adding more flags to the type.

Treating that node as an unexpanded pack would also benefit our diagnostics, see #121044

ilya-biryukov replied:
`SubstTemplateTypeParmPackType` needs a (variadic) template parameter and we don't have one.

We could go in the direction of making the template parameter in `SubstTemplateTypeParmPackType` optional (e.g. it's just null), but I believe it's a bad idea because that would require a lot of care to all places where `SubstTemplateTypeParmPackType` is used to make sure they handle the cases with nulls properly and having some static typing here should make these problems avoidable by design.

Another alternative that I imagine is to remove `SubstTemplateTypeParmPackType`, have only `SubstPackType` as a concrete type for both cases (builtin packs and variadic template parameters) and have different sugar on top: 
- for cases of delayed pack expansion (i.e. `SubstTemplateTypeParmPackType` now) we could use `SubstTemplateTypeParmType` that matches the sugar we want here precisely ("this delayed pack is coming from substitution of this template parameter")
- for cases with builtins producing packs, we add `TemplateSpecializationType` on top like we do today. 

I did not pursue this only because there was a curious use of `SubstTemplateTypeParmPackType` in template deduction that I didn't want to dig into to keep myself focused on the task at hand.
It feels like a good alternative, actually, but I would really want to pursue it separately to avoid changing behavior where we do not actually have these new builtins with this change.

mizvekov replied:
I haven't taken the time to review this in-depth yet, sorry but I have been busy with other things.

Taking a step back a little bit, I am a bit surprised the changes for implementing these builtins became so involved, with new AST nodes and changes to TreeTransform and all of that.

I would have expected the existing builtin template infrastructure to handle this quite well, and the implementation would only need to touch around `checkBuiltinTemplateIdType` in `SemaTemplate.cpp`.

So what happened, is this trying to go for more performance than what that simpler implementation would have provided?

If that's the case, can we try the simpler implementation first, and then go more complicated if needed, with numbers to back it up?

zyn0217 replied:
> So what happened, is this trying to go for more performance than what that simpler implementation would have provided?

As far as I can understand (@cor3ntin told me some contexts, thanks!), this builtin isn't just only used to deduplicate types, `dedup<T...>` also introduces a new form of pack and can be used anywhere like a regular parameter pack.

ilya-biryukov replied:
Thanks @zyn0217 for capturing the discussion.

FWIW, I have also been in favor of a more direct implementation that is simpler (see #139730), but I ultimately lost this battle (which is fine, as long as this implementation does have a chance to contribute to a new feature in the future). The downside is that my response times got really long as I couldn't find the time to dive deeper into the more complex implementation we need here, but that's on me (and sorry about that).

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -5054,95 +5089,128 @@ bool TreeTransform<Derived>::TransformTemplateArguments(
+    bool SawPackTypes = llvm::any_of(Unexpanded, [](UnexpandedParameterPack P) {
+      return P.first.dyn_cast<const SubstBuiltinTemplatePackType *>();
+    });
+    if (!SawPackTypes) {
+      Outputs.addArgument(In);
+      return false;
+    }
```
I added an assert to line 5128 and nothing fired with the test. Probably dead code?

ilya-biryukov commented:
I have started picking this back up after a break, sorry for the delays. I will address the comments in the coming days and send an updated version. #139730 also didn't really pick up, so I'm going to abandon the simpler attempt and double down on this PR.

cor3ntin commented:
@ilya-biryukov Great to hear. Please ping me when you are ready, I'll do my best to prioritize this

ilya-biryukov commented:
@cor3ntin please take a look at the new version.

I have split the sort builtin into #152118 (currently in draft, will update it as soon as this lands).
Other than that, the new implementation is a tad simpler because instead of using recursion it merely transforms the pattern, so the code is more clear.
I have also added support for bases, but ultimately decided to keep the other places that do expansion out of the initial change and instead produce an error there. Most of them should be easy to add later, but it does not look like any of those are blocking.

Eager to hear your thoughts on this.
I tried to address all comments that were left before, but there were quite a few so please reiterate them if I forgot something.

ilya-biryukov committed:
Use wider int type to avoid truncating builting

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -3736,12 +3735,9 @@ TemplateDeclInstantiator::VisitTemplateTemplateParmDecl(
-                                                NumExpansions))
+    if (SemaRef.CheckParameterPacksForExpansion(
+            D->getLocation(), TempParams->getSourceRange(), Unexpanded,
+            TemplateArgs, true, Expand, RetainExpansion, NumExpansions))
```
and here

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -4014,8 +4010,8 @@ Decl *TemplateDeclInstantiator::instantiateUnresolvedUsingDecl(
-          D->getEllipsisLoc(), D->getSourceRange(), Unexpanded, TemplateArgs,
-            Expand, RetainExpansion, NumExpansions))
+            D->getEllipsisLoc(), D->getSourceRange(), Unexpanded, TemplateArgs,
+            true, Expand, RetainExpansion, NumExpansions))
```
here

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -6414,11 +6410,9 @@ Sema::InstantiateMemInitializers(CXXConstructorDecl *New,
-                                          RetainExpansion,
-                                          NumExpansions)) {
+                                          BaseTL.getSourceRange(), Unexpanded,
+                                          TemplateArgs, true, ShouldExpand,
```
last time?

erichkeane commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -777,12 +844,32 @@ bool Sema::CheckParameterPacksForExpansion(
+                                             : EllipsisLoc,
+           diag::err_unsupported_builtin_template_pack_position)
+          << TST->getTemplateName();
+      return true;
```
Just do `return Diag` here instead!

erichkeane reviewed (commented):
I spent less time with TreeTransform.h than I wish I did, but in general I have only nits.  I see a few more of the 'add a comment to bool arguments' in TreeTransform that I didn't copy on.

In general, I think I'm down to just nits.  IF they are dealt with, feel free to accept an approval from the others while I'm gone (until Aug 14th).

zyn0217 commented on clang/test/SemaCXX/pr100095.cpp:
```diff
@@ -1,5 +1,4 @@
 // RUN: %clang_cc1 -fsyntax-only -std=c++11 %s
-// XFAIL: asserts
```
Is it fixed? We may want a release entry

zyn0217 commented on clang/include/clang/AST/Type.h:
```diff
@@ -6666,7 +6720,7 @@ class SubstTemplateTypeParmType final
 /// arguments), this type will be replaced with the \c SubstTemplateTypeParmType
 /// at the current pack substitution index.
-class SubstTemplateTypeParmPackType : public Type, public llvm::FoldingSetNode {
+class SubstTemplateTypeParmPackType : public SubstPackType {
```
I think you want to remove the Arguments member around line 6728. They are now in the SubstPackType, right?

zyn0217 commented on clang/include/clang/AST/Type.h:
```diff
@@ -6654,6 +6658,56 @@ class SubstTemplateTypeParmType final
+
+  /// A pointer to the set of template arguments that this
+  /// parameter pack is instantiated with.
+  const TemplateArgument *Arguments;
```
> Are they part of the containing declaration?

I think so, at least it's the case for SubstTemplateTypeParmPackType.

zyn0217 commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -228,7 +228,9 @@ void threadSafetyCleanup(BeforeSet *Cache);
 // FIXME: No way to easily map from TemplateTypeParmTypes to
 // TemplateTypeParmDecls, so we have this horrible PointerUnion.
-typedef std::pair<llvm::PointerUnion<const TemplateTypeParmType *, NamedDecl *>,
+typedef std::pair<llvm::PointerUnion<const TemplateTypeParmType *, NamedDecl *,
+                                     const TemplateSpecializationType *,
+                                     const SubstBuiltinTemplatePackType *>,
```
We're using the last 2 bits of PointerUnion on 32 bit systems, so it might be much harder if we want to add a new type of pack in future.

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -5154,6 +5177,84 @@ bool TreeTransform<Derived>::TransformTemplateArguments(
+  if (!OutPattern.getArgument().containsUnexpandedParameterPack())
+    return false;
+
+  // Some packs will learn their length after substitution.
+  // We may need to request their expansion.
```
Can we have more explanations/examples?

zyn0217 commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -777,12 +844,32 @@ bool Sema::CheckParameterPacksForExpansion(
+                   ParmPack.first
+                       .dyn_cast<const TemplateSpecializationType *>()) {
+      assert(isPackProducingBuiltinTemplateName(TST->getTemplateName()));
+      // Delay expansion, substitution is required to know the size.
+      ShouldExpand = false;
+      if (!FailOnPackProducingTemplates)
+        continue;
```
Instead of an explicit flag, I wonder if we can introspect the situation where we don't need the diagnostics. E.g. can we check the dependency flags of .. template arguments?

cor3ntin replied:
Can we check that we are in a dependent context instead of an extra bool? Not sure it would work but it's worth exploring

zyn0217 commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -9,18 +9,25 @@
 #include "clang/Sema/Template.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/Support/SaveAndRestore.h"
+#include <memory>
```
What is it used for?
Can you remove unnecessary auto-includes here and there?

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -3495,47 +3609,37 @@ Sema::SubstBaseSpecifiers(CXXRecordDecl *Instantiation,
-                                          TemplateArgs, ShouldExpand,
-                                          RetainExpansion,
-                                          NumExpansions)) {
+      UnexpandedInfo Info;
+      if (PreparePackForExpansion(*this, Base, TemplateArgs, BaseTypeLoc,
+                                  Info)) {
```
Hum I'm thinking if we can move PreparePackForExpansion into CheckParameterPacksForExpansion, thereby we can get rid of the duplication and make it easier for more type supports in future.

But I don't think it's required for now, we may want a FIXME around PreparePackForExpansion :)

zyn0217 reviewed (commented):
Thanks, I left some comments.

Also ping @mizvekov @cor3ntin if they have different opinions. Otherwise feel free to merge (after resolving comments)

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6074,6 +6074,13 @@ def warn_cxx23_pack_indexing : Warning<
 def err_pack_outside_template : Error<
   "pack declaration outside of template">;
 
+def err_builtin_pack_outside_template
+    : Error<"builtin returning packs used outside of template">;
```
Maybe just '%0 cannot be used outside of template'

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -3478,6 +3529,69 @@ bool Sema::SubstDefaultArgument(
+    Info.Expand = false;
+    Info.RetainExpansion = false;
+    Info.NumExpansions = std::nullopt;
+    return S.CheckParameterPacksForExpansion(
+        BaseEllipsisLoc, BaseSourceRange, Unexpanded, TemplateArgs, false,
+        Info.Expand, Info.RetainExpansion, Info.NumExpansions);
+  };
```
Need a comment here too :)

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -30,22 +30,30 @@
 #include "clang/AST/StmtOpenACC.h"
 #include "clang/AST/StmtOpenMP.h"
 #include "clang/AST/StmtSYCL.h"
+#include "clang/AST/TemplateBase.h"
+#include "clang/AST/Type.h"
+#include "clang/AST/TypeLoc.h"
 #include "clang/Basic/DiagnosticParse.h"
 #include "clang/Basic/OpenMPKinds.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Basic/UnsignedOrNone.h"
 #include "clang/Sema/Designator.h"
 #include "clang/Sema/EnterExpressionEvaluationContext.h"
 #include "clang/Sema/Lookup.h"
 #include "clang/Sema/Ownership.h"
 #include "clang/Sema/ParsedTemplate.h"
 #include "clang/Sema/ScopeInfo.h"
+#include "clang/Sema/Sema.h"
 #include "clang/Sema/SemaDiagnostic.h"
 #include "clang/Sema/SemaInternal.h"
 #include "clang/Sema/SemaObjC.h"
 #include "clang/Sema/SemaOpenACC.h"
 #include "clang/Sema/SemaOpenMP.h"
 #include "clang/Sema/SemaPseudoObject.h"
 #include "clang/Sema/SemaSYCL.h"
+#include "clang/Sema/Template.h"
 #include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
```
I suspect none of these changes are needed

ilya-biryukov replied:
Removed the #includes

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -4477,11 +4531,9 @@ bool TreeTransform<Derived>::TransformExprs(Expr *const *Inputs,
-                                               Unexpanded,
-                                               Expand, RetainExpansion,
-                                               NumExpansions))
+      if (getDerived().TryExpandParameterPacks(
+              Expansion->getEllipsisLoc(), Pattern->getSourceRange(),
+              Unexpanded, true, Expand, RetainExpansion, NumExpansions))
```
we would need a `/*FailOnPackProducingTemplates=/*` comment here

ilya-biryukov replied:
Done.

cor3ntin commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -5694,6 +5694,35 @@ ASTContext::getSubstTemplateTypeParmPackType(Decl *AssociatedDecl,
+    return QualType(T, 0);
+
+  QualType Canon;
+  {
+    TemplateArgument CanonArgPack = getCanonicalTemplateArgument(ArgPack);
+    if (!CanonArgPack.structurallyEquals(ArgPack))
+      Canon = getSubstBuiltinTemplatePack(CanonArgPack);
+  }
```
I'm not sure the braces are necessary here, letting CanonArgPack survive until the end of the function should not matter

cor3ntin commented on clang/lib/AST/DeclTemplate.cpp:
```diff
@@ -275,6 +275,17 @@ void *allocateDefaultArgStorageChain(const ASTContext &C) {
 
+bool isPackProducingBuiltinTemplate(const TemplateDecl *D) {
+  auto *BD = llvm::dyn_cast<BuiltinTemplateDecl>(D);
+  return BD && BD->getBuiltinTemplateKind() == clang::BTK__builtin_dedup_pack;
```
Can we put that logic in BuiltinTemplateDecl ?

cor3ntin commented on clang/lib/AST/Type.cpp:
```diff
@@ -4522,6 +4561,12 @@ QualType TemplateSpecializationType::getAliasedType() const {
+bool clang::TemplateSpecializationType::isSugared() const {
+  return !isDependentType() || isCurrentInstantiation() || isTypeAlias() ||
+         (isPackProducingBuiltinTemplateName(Template) &&
+          llvm::isa<SubstBuiltinTemplatePackType>(*getCanonicalTypeInternal()));
```
```suggestion
          isa<SubstBuiltinTemplatePackType>(*getCanonicalTypeInternal()));
```

cor3ntin commented on clang/lib/Parse/ParseTemplate.cpp:
```diff
@@ -1305,6 +1307,17 @@ ParsedTemplateArgument Parser::ParseTemplateTemplateArgument() {
+  // We do not allow to reference builtin templates that produce multiple
+  // values, they would not have a well-defined semantics outside template
+  // arguments.
+  if (auto *T = !Result.isInvalid()
+                    ? Result.getAsTemplate().get().getAsTemplateDecl()
+                    : nullptr;
+      T && isPackProducingBuiltinTemplate(T)) {
+    Actions.diagnoseMissingTemplateArguments(Result.getAsTemplate().get(),
+                                             Result.getLocation());
+  }
```
I would prefer to either: do that check in Sema or introduce a new Parser diagnostic.
Calling Action.diagnoseMissingTemplateArguments for that does seem like a layering violation

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -17,30 +17,44 @@
 #include "clang/AST/Expr.h"
 #include "clang/AST/ExprCXX.h"
+#include "clang/AST/Mangle.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/AST/TemplateBase.h"
 #include "clang/AST/TemplateName.h"
 #include "clang/AST/Type.h"
+#include "clang/AST/TypeLoc.h"
+#include "clang/AST/TypeOrdering.h"
 #include "clang/AST/TypeVisitor.h"
 #include "clang/Basic/Builtins.h"
 #include "clang/Basic/DiagnosticSema.h"
 #include "clang/Basic/LangOptions.h"
 #include "clang/Basic/PartialDiagnostic.h"
 #include "clang/Basic/SourceLocation.h"
 #include "clang/Basic/TargetInfo.h"
+#include "clang/Basic/UnsignedOrNone.h"
 #include "clang/Sema/DeclSpec.h"
 #include "clang/Sema/EnterExpressionEvaluationContext.h"
 #include "clang/Sema/Initialization.h"
 #include "clang/Sema/Lookup.h"
 #include "clang/Sema/Overload.h"
 #include "clang/Sema/ParsedTemplate.h"
 #include "clang/Sema/Scope.h"
+#include "clang/Sema/Sema.h"
 #include "clang/Sema/SemaCUDA.h"
 #include "clang/Sema/SemaInternal.h"
 #include "clang/Sema/Template.h"
 #include "clang/Sema/TemplateDeduction.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallBitVector.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/Support/Casting.h"
 #include "llvm/Support/SaveAndRestore.h"
+#include "llvm/Support/raw_ostream.h"
```
More unneeded header changes I'd like to see reverted

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -316,6 +330,13 @@ TemplateNameKind Sema::isTemplateName(Scope *S,
     }
   }
 
+  if (isPackProducingBuiltinTemplateName(Template) &&
+      S->getTemplateParamParent() == nullptr) {
+    Diag(Name.getBeginLoc(), diag::err_builtin_pack_outside_template);
+    // Recover by returning the template, even though we would never be able to
+    // substitute it.
+  }
```
```suggestion
  if (isPackProducingBuiltinTemplateName(Template) &&
      S->getTemplateParamParent() == nullptr)
    Diag(Name.getBeginLoc(), diag::err_builtin_pack_outside_template);
    // Recover by returning the template, even though we would never be able to
    // substitute it.
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -36,6 +40,7 @@
 #include "clang/Sema/TemplateDeduction.h"
 #include "clang/Sema/TemplateInstCallback.h"
 #include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Casting.h"
```
Ditto

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6074,6 +6074,13 @@ def warn_cxx23_pack_indexing : Warning<
+def err_builtin_pack_outside_template
+    : Error<"builtin returning packs used outside of template">;
+
+def err_unsupported_builtin_template_pack_position
+    : Error<"expansions of %0 are not supported here. Only expansions in "
+            "template arguments and class bases are supported">;
```
```suggestion
def err_unsupported_builtin_template_pack_expansion
    : Error<"expansions of %0 are not supported here. Only expansions in "
            "template arguments and class bases are supported">;
```

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -5857,6 +5900,29 @@ bool Sema::CheckTemplateArgumentList(
       }
     }
 
+    // Check for builtins producing template packs at this position, we do not
+    // support them yet.
```
```suggestion
    // Check for builtins producing template packs in this context, we do not
    // support them yet.
```

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -5857,6 +5900,29 @@ bool Sema::CheckTemplateArgumentList(
+        if (!TST)
+          continue;
+        assert(isPackProducingBuiltinTemplateName(TST->getTemplateName()));
+        // It is not yet supported in many positions.
```
```suggestion
        // expanding a built-in pack in this context is not yet supported.
```

Please review any other use of "position" to say context instead

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -935,15 +938,18 @@ class PackDeductionScope {
       for (unsigned I = 0, N = Unexpanded.size(); I != N; ++I) {
         unsigned Depth, Index;
-        std::tie(Depth, Index) = getDepthAndIndex(Unexpanded[I]);
+        if (auto DI = getDepthAndIndex(Unexpanded[I])) {
+          std::tie(Depth, Index) = *DI;
+        } else {
+          continue;
+        }
```
Per our styles guides you should remove the braces

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1510,6 +1517,21 @@ namespace {
+      MultiLevelTemplateArgumentList New;
+      New.addOuterRetainedLevels(this->TemplateArgs.getNumLevels());
+
+      MultiLevelTemplateArgumentList Old =
+          const_cast<MultiLevelTemplateArgumentList &>(this->TemplateArgs);
```
Do you need that first const_cast?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -3478,6 +3529,69 @@ bool Sema::SubstDefaultArgument(
   return false;
 }
 
+// See TreeTransform::PreparePackForExpansion for the relevant comment.
+// This function implements the same concept for base specifiers.
```
Is there anyway we could reduce the code duplication here?

ilya-biryukov committed:
remove includes added by Clangd

ilya-biryukov committed:
Pass the builtin name in the error message

ilya-biryukov committed:
Rename err_unsupported_builtin_template_pack_position to _expansion

ilya-biryukov committed:
Use isa<> with multiple template arguments

ilya-biryukov committed:
Remove unnecessary braces

ilya-biryukov committed:
Move helper into BuiltinTemplateDecl

ilya-biryukov committed:
Update mangled name

ilya-biryukov committed:
Simplify an assertion

ilya-biryukov committed:
Remove redundant llvm:: prefix

ilya-biryukov committed:
Move check for the presence of template argument from parser into sema

ilya-biryukov committed:
Add a comment with parameter name

ilya-biryukov committed:
use if instead of continue

ilya-biryukov committed:
Revert added #includes

ilya-biryukov committed:
Remove braces

ilya-biryukov committed:
Updated comments to use context instead of position

ilya-biryukov committed:
Remove redundant braces

ilya-biryukov committed:
remove redundant includes

ilya-biryukov committed:
Added a FIXME about reducing code duplication

ilya-biryukov committed:
Add a comment from the duplicated version

ilya-biryukov committed:
Remove #includes

ilya-biryukov committed:
Return Diag()

ilya-biryukov committed:
Add comment, removed #includes

ilya-biryukov committed:
Fix a typo, add a comment

ilya-biryukov committed:
Add Template.h, it is needed

ilya-biryukov committed:
Added requested tests

ilya-biryukov commented on clang/test/Import/builtin-template/test.cpp:
```diff
@@ -20,4 +22,10 @@ void expr() {
+
+#ifdef DEDUP
+  static_assert(__is_same(TypePackDedup<TypeList>, TypeList<>), "");
+  static_assert(__is_same(TypePackDedup<TypeList, int, double, int>, TypeList<int, double>), "");
```
I have added a few more tests, PTAL to make sure they cover what you intended.
There is also `dedup-types-builtin.cpp` test that does a much more exhaustive test of the actual results for the builtin rather than `-import` logic. The import is actually pretty bulletproof now that TableGen is used for it, so I am not too worried about missing anything here.

To give some context why this test is here in the first place. The original change predates the TableGen and this was needed to test actual failures I have encountered >1 year ago when I started poking at this builtin.
It is not strictly necessary now, but I also felt it would not hurt and kept it from the previous versions.

ilya-biryukov committed:
Remove more unused #includes

ilya-biryukov committed:
Fix test crashes

mizvekov commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1700,6 +1717,21 @@ namespace {
       return inherited::TransformTemplateArgument(Input, Output, Uneval);
     }
 
+    using TreeTransform::TransformTemplateSpecializationType;
+    QualType
+    TransformTemplateSpecializationType(TypeLocBuilder &TLB,
+                                        TemplateSpecializationTypeLoc TL) {
+      auto *T = TL.getTypePtr();
+      if (!getSema().ArgPackSubstIndex || !T->isSugared() ||
+          !isPackProducingBuiltinTemplateName(T->getTemplateName()))
+        return TreeTransform::TransformTemplateSpecializationType(TLB, TL);
+      // Look through sugar to get to the SubstBuiltinTemplatePackType that we
+      // need to substitute into.
+      QualType R = TransformType(T->desugar());
+      TLB.pushTrivial(getSema().getASTContext(), R, TL.getBeginLoc());
+      return R;
+    }
```
Why is this simply dropping the template specialization type from the resulting transformed AST?

That's not something we do anywhere else, and this can create problems for people analyzing the AST, ie the underlying template specialization type pops out of nowhere and it's not clear what it was produced from.

The original logic of classifying all builtin templates as type aliases also would have helped here, the original builtin TST would simply become a normal type alias.

ilya-biryukov replied:
> "with the pack indexing stuff that we currently have, right?"

I though we do not have anything that fits right now. The options I was aware of:
- `SubstBuiltinTemplatePackType` (and `SubstPackType` added in this patch): represent the whole pack, not the n-th element
- `SubstTemplateTypeParmType`: represents the n-th element, but only for a template parameter (and we have a builtin with template arguments applied to it)

I believe you refer to `PackIndexingType`? I have previously overlooked it. It seems to fit almost well, but because it assumes a particular C++26 syntax it's still not quite great:
- it needs some things we do not have in that situation, i.e. ellipsis location and expression for the index (rather than just a numeric index)
- it will be handled and printed as this C++26 syntax even though we did not have it in the source code, potentially leading to confusion.

But more importantly, when reading the code it struck me that we can end up having `O(n^2)` compile times where `n` is the size of the template pack in certain cituations. I filed #153431 to discuss this further. Given that this builtin is solely done for compile times, I want to make sure we avoid this behavior. Although I feel we should also make sure pack indexing is as efficient as it can be.

mizvekov replied:
Yeah, I meant PackIndexType, and sure if the performance there is not great, we can leave this change for later.

But I'd think you be able to build a new AST Node which is basically a stripped down base class of PackIndexType which would be a good fit for this, and that's not as much work as building something from scratch.

ilya-biryukov committed:
Revert unintentional change

ilya-biryukov committed:
Add a comment about SUBSTBUILTINPACK

ilya-biryukov committed:
rename Dep to Dependence

ilya-biryukov committed:
Add a clarifying comment and a fixme for sugar

ilya-biryukov committed:
Merge remote-tracking branch 'origin/main' into dedup

ilya-biryukov committed:
Fix compilation after merge

ilya-biryukov committed:
Format the code

cor3ntin commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -5885,6 +5885,33 @@ ASTContext::getSubstTemplateTypeParmPackType(Decl *AssociatedDecl,
+QualType
+ASTContext::getSubstBuiltinTemplatePack(const TemplateArgument &ArgPack) {
+#ifndef NDEBUG
+  for (const auto &P : ArgPack.pack_elements())
+    assert(P.getKind() == TemplateArgument::Type && "Pack contains a non-type");
```
nit: can we use all_of ?

cor3ntin commented on clang/test/SemaTemplate/dedup-types-builtin.cpp:
```diff
@@ -0,0 +1,226 @@
+
+template <class ...T>
+struct MemInitializers : T... {
+  MemInitializers() : __builtin_dedup_pack<T...>()... {} // expected-error 2 {{expansions of '__builtin_dedup_pack' are not supported here.}}
```
```suggestion
  MemInitializers() : __builtin_dedup_pack<T...>()... {} // expected-error 2{{expansions of '__builtin_dedup_pack' are not supported here.}}
```

We don't put spaces there usually (same in the rest of the file)

cor3ntin reviewed (commented):
Modulo nits, I think this is almost ready to be landed. I hope the few fixme are investigated in follow ups.

Can you add a release note and some documentation in LanguageExtensions.rst, in the [Type Trait Primitives](https://clang.llvm.org/docs/LanguageExtensions.html#type-trait-primitives) section?

I hope @mizvekov @zyn0217 and @erichkeane can do a last pass on this PR

ilya-biryukov committed:
Add a release note

ilya-biryukov committed:
Add documentation about the new builtin to LanguageExtensions.rst

ilya-biryukov committed:
replace loop with all_of

ilya-biryukov committed:
Remove unnecessary spaces

ilya-biryukov committed:
Merge remote-tracking branch 'origin/main' into dedup

ilya-biryukov committed:
add newline to fix documentation build errors

ilya-biryukov committed:
Remove unnecessary check for TemplateName::DeducedTemplate

ilya-biryukov committed:
Merge remote-tracking branch 'origin/main' into dedup

cor3ntin reviewed (approved):
LGTM
Please give it another ~24h  in case someone else has anything to add.

Thanks for the huge amount of work done here, this will be super useful.

zyn0217 reviewed (approved):
Sorry for not responding in time, I'm spending time in #141776 and there are many things to be done there.

Overall, this LGTM. Thanks again for improving template packs!

mizvekov commented on clang/lib/AST/DeclTemplate.cpp:
```diff
@@ -1598,6 +1599,16 @@ BuiltinTemplateDecl::BuiltinTemplateDecl(const ASTContext &C, DeclContext *DC,
+}
+
+bool clang::isPackProducingBuiltinTemplateName(TemplateName N) {
+  auto *T = dyn_cast_or_null<BuiltinTemplateDecl>(
+      N.getTemplateDeclAndDefaultArgs().first);
```
Small nit, slightly cleaner:
```suggestion
  auto *T = dyn_cast_or_null<BuiltinTemplateDecl>(
      N.getAsTemplateDecl(/*IgnoreDeduced=*/true));
```

mizvekov replied:
FYI the performance of this is also better, since we won't build the defaulted argument list.

ilya-biryukov replied:
Thanks for spotting this! Done.

mizvekov reviewed (approved):
LGTM as well, thanks for working on this!

ilya-biryukov committed:
Use `getAsTemplateDecl()`

ilya-biryukov committed:
Merge remote-tracking branch 'origin/main' into dedup

kazutakahirata commented:
@ilya-biryukov After this PR, I'm getting:

```
lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4148:11: error: enumeration value 'SubstBuiltinTemplatePack' not handled in switch [-Werror,-Wswitch]
 4148 |   switch (qual_type->getTypeClass()) {
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:4852:11: error: enumeration value 'SubstBuiltinTemplatePack' not handled in switch [-Werror,-Wswitch]
 4852 |   switch (qual_type->getTypeClass()) {
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:5153:11: error: enumeration value 'SubstBuiltinTemplatePack' not handled in switch [-Werror,-Wswitch]
 5153 |   switch (qual_type->getTypeClass()) {
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

Is there any way you could take look?

kazutakahirata commented:
@ilya-biryukov I've addressed these warnings with https://github.com/llvm/llvm-project/pull/154606.

dyung commented on clang/test/SemaTemplate/dedup-types-builtin.cpp:
```diff
@@ -0,0 +1,226 @@
+template <class ...T>
+struct LambdaInitCaptures {
+  static constexpr int test() {
+    [...foos=__builtin_dedup_pack<T...>()]{}; // expected-warning {{initialized lambda pack captures are a C++20 extension}} \
```
@ilya-biryukov this warning is not present in a target that defaults to C++20 which causes this test to fail. Can the test be adjusted to work with both C++20 and non-C++20 defaulting targets?

ilya-biryukov replied:
Addressed in 26d4e56be81adbaa8aeed7f5cde39b779d959fc3, thanks for bringing this up.

ilya-biryukov commented:
> @ilya-biryukov I've addressed these warnings with #154606.

Thanks a lot! And sorry for not spotting this earlier, I do not have LLDB build configured locally. It would be great to add this to presubmits at some point in order to spot those problems.
